### ISSUE TITLE
[CLOSED] Tests for checkpoint logic + some fixes to checkpointing

### ISSUE THREAD

**Issue by jeswan** [nyu-mll/jiant-v1-legacy#692]
<a href="https://github.com/pruksmhc"><img src="https://avatars3.githubusercontent.com/u/10094008?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [pruksmhc](https://github.com/pruksmhc)**
_Wednesday May 22, 2019 at 10:34 GMT_
_Originally opened as https://github.com/nyu-mll/jiant/pull/692_

----

This PR is dedicated to creating tests for checkpointing and fixing some checkpointing errors. 
The general flow is the below:
-We save the best metric thus far. For pretraining, we do multitask training, and save during each epoch the model states (both the sentence encoder and the task-specific modules).
-For finetuning, we must also save task-specific versions of the sentence encoder. If we don't allow pretrained weights for the task-specific modules, we do not load the parameter weights for a module if the sentence encoder has not been trained on a specific task yet. 

Changes:
-Making the code less specific to target_train and adding asserts to make sure that the checkpoints loaded are the best ones rather than the most recent. For example, removing redundant model_target_state_best.th checkpoint. 
-Adding the demo.sh run to .circleCi build since running the script from tests seems odd to make sure that the output of demo.sh is the same. 

----

_**[pruksmhc](https://github.com/pruksmhc)** included the following code: https://github.com/nyu-mll/jiant/pull/692/commits_
